### DISCUSSION TITLE
No bluetooth output option after successfully adding a bluetooth speaker

### DISCUSSION THREAD

**Discussion by davidasbury** [rern/rAudio#872] [Bug Reports]
@rern Nice work on this! I'm really digging your player.

However, I can't seem to figure out how to change the output from On-board Headphones to a connected Bluetooth speaker. The bluetooth handshake is successful, but there's no option to change the device under Player -> Output.

<img src="https://user-images.githubusercontent.com/6776067/161876860-0c819bd1-eb0c-4dbb-b501-75574f9d1623.jpg" width=300>

also tried to manually overwrite asound.conf and mpd.conf, but didn't have any luck. 

Running version 202203271 on a rPi 3B Rev 1.2.

Any help would be greatly appreciated!

**Comment by ceejayemm** [comment]
@davidasbury 

It took me a while to get my head around this too. However provided your BT Speakers are paired with rAudio-1 and turned on then you don't need to do anything else other than select whatever you want to listen to from your Library and hit Play on the Playback screen.  There is no further config needed nor manual changes to any files.

Try it !

My setup is RPi 3A, Soundcore 2 BT Speakers and rAudio-1 202203271 - it just works

Chris

**Reply by davidasbury** [reply]
@ceejayemm 

Thanks, but unfortunately that's not working for me. The speaker is working if I connect the headphone jack, but not through bluetooth. 

My mpd.conf still shows audio_output set as "On-Board Headphones". Is this similar to what you see in your config?

<img src="https://user-images.githubusercontent.com/6776067/162023125-0719908d-4049-4c93-ac58-eacef25bed40.jpg" width=300>

```rAudio:~  #  cat /etc/mpd.conf
bind_to_address        "0.0.0.0"
db_file                "/srv/http/data/mpd/mpd.db"
music_directory        "/mnt/MPD"
playlist_directory     "/srv/http/data/playlists"
replaygain             "off"
state_file             "/var/lib/mpd/mpdstate"
user                   "mpd"

input {
	plugin         "curl"
}

decoder {
	plugin         "ffmpeg"
	enabled        "no"
}

resampler {
	plugin         "soxr"
	quality        "very high"
}

audio_output {
	name           "On-board Headphones"
	device         "hw:0,0"
	type           "alsa"
	auto_resample  "no"
	auto_format    "no"
	mixer_type     "none"
}
```

**Comment by ceejayemm** [comment]
@davidasbury 

Sorry for asking what might be stupid questions but:

- on the System page - is the Bluetooth option is turned on ?
- on the Networks / Bluetooth options - does it show your speakers (in my SoundCore 2) with a green dot next to them to confirm they are paired and connected
- Under Player / Output - does Bluetooth show your BT device (in my case SoundCore 2 - A2DP) is operational

I have NOT had to do anything else to make this work.

My mpd.conf shows:

mpd 0.23.5-1
bind_to_address        "0.0.0.0"
#db_file                "/srv/http/data/mpd/mpd.db"
music_directory        "/mnt/MPD"
playlist_directory     "/var/lib/mpd/playlists"
replaygain             "off"
state_file             "/var/lib/mpd/mpdstate"
user                   "mpd"

database {
    plugin "proxy"
    host "192.168.86.31"
}

input {
	plugin         "curl"
}

decoder {
	plugin         "ffmpeg"
	enabled        "yes"
}

resampler {
	plugin         "soxr"
	quality        "very high"
}

audio_output {
	name           "SoundCore 2"
	device         "bluealsa"
	type           "alsa"
	mixer_type     "hardware"
}

● mpd.service - Music Player Daemon
     Loaded: loaded (/usr/lib/systemd/system/mpd.service; disabled; vendor preset: disabled)
    Drop-In: /etc/systemd/system/mpd.service.d
             └─override.conf
     Active: active (running) since Wed 2022-04-06 17:38:49 BST; 20min ago
       Docs: man:mpd(1)
             man:mpd.conf(5)
   Main PID: 62132 (mpd)
      Tasks: 6 (limit: 442)
        CPU: 25.819s
     CGroup: /system.slice/mpd.service
             └─62132 /usr/bin/mpd --no-daemon

Apr 06 17:38:51 rAudio-1 taskset[62132]: D: bluealsa-pcm.c:489: /org/bluealsa/hci0/dev_08_EB_ED_7F_C8_B7/a2dpsrc/sink: Initializing SW
Apr 06 17:38:51 rAudio-1 taskset[62132]: output: Failed to open mixer for 'SoundCore 2': no such mixer control: PCM
Apr 06 17:38:51 rAudio-1 taskset[62132]: D: bluealsa-pcm.c:345: /org/bluealsa/hci0/dev_08_EB_ED_7F_C8_B7/a2dpsrc/sink: Starting
Apr 06 17:38:51 rAudio-1 taskset[62132]: D: bluealsa-pcm.c:212: /org/bluealsa/hci0/dev_08_EB_ED_7F_C8_B7/a2dpsrc/sink: Starting IO loop: 23
Apr 06 17:48:27 rAudio-1 taskset[62132]: player: played "http://stream.live.vc.bbcmedia.co.uk/bbc_radio_two"
Apr 06 17:48:27 rAudio-1 taskset[62132]: D: bluealsa-pcm.c:381: /org/bluealsa/hci0/dev_08_EB_ED_7F_C8_B7/a2dpsrc/sink: Stopping
Apr 06 17:48:27 rAudio-1 taskset[62132]: D: bluealsa-pcm.c:147: /org/bluealsa/hci0/dev_08_EB_ED_7F_C8_B7/a2dpsrc/sink: IO thread cleanup
Apr 06 17:48:27 rAudio-1 taskset[62132]: D: bluealsa-pcm.c:381: /org/bluealsa/hci0/dev_08_EB_ED_7F_C8_B7/a2dpsrc/sink: Stopping
Apr 06 17:48:27 rAudio-1 taskset[62132]: D: bluealsa-pcm.c:481: /org/bluealsa/hci0/dev_08_EB_ED_7F_C8_B7/a2dpsrc/sink: Freeing HW
Apr 06 17:48:27 rAudio-1 taskset[62132]: D: bluealsa-pcm.c:431: /org/bluealsa/hci0/dev_08_EB_ED_7F_C8_B7/a2dpsrc/sink: Closing

The Output config is:

# cat /etc/asound.conf
defaults.pcm.card 0
defaults.ctl.card 0
pcm.bluealsa {
	type plug
	slave.pcm {
		type bluealsa
		device 00:00:00:00:00:00
		profile "a2dp"
	}
}

# aplay -l | grep ^card
card 0: Headphones [bcm2835 Headphones], device 0: bcm2835 Headphones [bcm2835 Headphones]

# amixer scontrols
Simple mixer control 'Headphone',0

# bluealsa-aplay -L
bluealsa:SRV=org.bluealsa,DEV=08:EB:ED:7F:C8:B7,PROFILE=a2dp
    SoundCore 2, audio-headset, playback
    A2DP (SBC): S16_LE 2 channels 44100 Hz


But I must re-emphasize that I have NOT made any manual changes to either of these configs.  If you get the three steeps at the start of this post correct then they get changed automatically.

Chris

**Reply by davidasbury** [reply]
@ceejayemm  There are no stupid questions, especially during troubleshooting!

>     on the System page - is the Bluetooth option is turned on ?
Yes
>     on the Networks / Bluetooth options - does it show your speakers (in my SoundCore 2) with a green dot next to them to confirm they are paired and connected
Yes
>     Under Player / Output - does Bluetooth show your BT device (in my case SoundCore 2 - A2DP) is operational
No

Under Player -> Output, the only device that's shown is On-Board Headphones and my asound.conf does not show any bluealsa device:

```
rAudio:~  #  cat /etc/asound.conf
defaults.pcm.card 0
defaults.ctl.card 0
```
But bluealsa-aplay does correctly list the device:

```
rAudio:~  #  bluealsa-aplay -L
bluealsa:SRV=org.bluealsa,DEV=<MAC address>,PROFILE=a2dp
    Asbury JBL, trusted audio-card, playback
    A2DP (SBC): S16_LE 2 channels 44100 Hz
```

**Comment by ceejayemm** [comment]
@davidasbury 

Sorry, can't think of anything else except - have you tried a reboot ? I thought @rern had been able to remove most cases of a reboot being needed but you never know.

Chris

**Reply by davidasbury** [reply]
@ceejayemm Yep, rebooted multiple times and disconnected/reconnected BT multiple times too. Still no love.
Thanks for the assist though!

**Comment by rern** [comment]
@davidasbury,
Let's see if there's `Audio Sink` in the output of `bluetoothctl info`. (while the Bluetooth is connected)

**Reply by davidasbury** [reply]
@rern 
Yep, it's there:
```
rAudio:~  #  bluetoothctl info
Device <MAC address> (public)
	Name: Asbury JBL
	Alias: Asbury JBL
	Class: 0x00240414
	Icon: audio-card
	Paired: yes
	Trusted: yes
	Blocked: no
	Connected: yes
	LegacyPairing: no
	UUID: Audio Sink                (0000110b-0000-1000-8000-00805f9b34fb)
	UUID: A/V Remote Control Target (0000110c-0000-1000-8000-00805f9b34fb)
	UUID: Advanced Audio Distribu.. (0000110d-0000-1000-8000-00805f9b34fb)
	UUID: A/V Remote Control        (0000110e-0000-1000-8000-00805f9b34fb)
```

**Comment by andyliuxp** [comment]
I have same problem with "davidasbury"

**Comment by davidasbury** [comment]
@rern Updated to 20220410 through the GUI (Addons -> rAudio -> Update) but still no change in output options. ``bluetoothctl`` shows audio sink is there, but it doesn't appear in ``asound.conf``

Any thoughts?

**Comment by rern** [comment]
Today's update should fix the issue.
After initial pairing, if it's still not show on Player page as an output, try power off the Bluetooth device then on again.

**Reply by davidasbury** [reply]
Looks like that did it! After installing 202204151 and rebooting, then "forgetting" previous bluetooth connection and reconnecting, bluetooth is working great!

Thanks so much!

**Reply by davidasbury** [reply]
@rern 
Looks like I may have spoken too soon... I'm seeing two issues:

1. Bluetooth connection is not persisting across reboots. Connecting BT speaker requires forgetting the connection, searching and then re-establishing after each reboot.
2. Random Bluetooth dropouts. The connection is fine for a while, but after a few minutes (between 30 seconds and 10 minutes) the sound stops and sometimes a blue toast message briefly appears in the lower left of the screen. The BT connection is still there (green dot in BT section of Networks settings), but no audio output until reboot and forget/reconnect BT speaker

Update: seems like the BT connection is more likely to drop if I'm accessing rAudio via an external browser or SSH. After doing a reboot and only using the rPi touchscreen to forget/re-connect the speaker, the connection has been steady for over an hour.

Another update: BT connection did eventually drop after about 2.5 hours, even with no intervention.

**Comment by rern** [comment]
Please try today's update.

**Reply by davidasbury** [reply]
@rern 
Looks like that did it! Sorry it took me a bit to test, but it ran for several hours without dropping bluetooth connection.
Thanks again!
